presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveRedundantLimit.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.LimitNode;
+
+import static io.prestosql.sql.planner.optimizations.QueryCardinalityUtil.isAtMost;
+import static io.prestosql.sql.planner.plan.Patterns.Limit.count;
+import static io.prestosql.sql.planner.plan.Patterns.limit;
+
+/**
+ *  Remove Limit node when the subplan is guaranteed to produce fewer rows than the limit
+ */
+public class RemoveRedundantLimit
+        implements Rule<LimitNode>
+{
+    private static final Pattern<LimitNode> PATTERN = limit().with(count().matching(limitCount -> limitCount != 0));
```

Comment:
What is wrong with `limitCount == 0`?

Reply:
Since both `RemoveRedundantLimit` and `EvaluateZeroLimit` rules are in the same optimizer , we don't want `LimitNode` with 0 count to be removed if its source is scalar as it might not fire `EvaluateZeroLimit` rule

Reply:
@kokosing , @martint Have merged both of the two rules